testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java

```diff
@@ -0,0 +1,584 @@
+            } finally {
+
+                RealmManager realmMgr = new RealmManager(currentSession);
+                RealmModel fooRealm = realmMgr.getRealm("foo");
+                UserModel user3 = currentSession.users().getUserByUsername("user3", fooRealm);
+
+                // Remove user3
+                new UserManager(currentSession).removeUser(fooRealm, user3);
+
+                // Cleanup
+                realmMgr = new RealmManager(currentSession);
+                realmMgr.removeRealm(realmMgr.getRealm("foo"));
+            }
+
+        });
+    }
+
+
+    @Test
+    @ModelTest
+    public void testOnUserRemoved(KeycloakSession session) {
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionUR) -> {
+            try {
+                int started = Time.currentTime();
+
+                RealmModel fooRealm = sessionUR.realms().createRealm("foo", "foo");
+                fooRealm.addClient("foo-app");
+                sessionUR.users().addUser(fooRealm, "user3");
+
+                UserSessionModel userSession = sessionUR.sessions().createUserSession(fooRealm, sessionUR.users().getUserByUsername("user3", fooRealm), "user3", "127.0.0.1", "form", true, null, null);
+                AuthenticatedClientSessionModel clientSession = createClientSession(sessionUR, fooRealm.getClientByClientId("foo-app"), userSession, "http://redirect", "state");
+
+                // Create offline session
+                fooRealm = sessionUR.realms().getRealm("foo");
+                userSession = sessionUR.sessions().getUserSession(fooRealm, userSession.getId());
+                createOfflineSessionIncludeClientSessions(sessionUR, userSession);
+
+                RealmManager realmMgr = new RealmManager(sessionUR);
+                fooRealm = realmMgr.getRealm("foo");
+                UserModel user3 = sessionUR.users().getUserByUsername("user3", fooRealm);
+
+                // Assert session was persisted with both clientSessions
+                UserSessionModel offlineSession = sessionUR.sessions().getOfflineUserSession(fooRealm, userSession.getId());
+                assertSession(offlineSession, user3, "127.0.0.1", started, started, "foo-app");
+
+
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            } finally {
+
+                RealmManager realmMgr = new RealmManager(sessionUR);
+                RealmModel fooRealm = realmMgr.getRealm("foo");
+                UserModel user3 = sessionUR.users().getUserByUsername("user3", fooRealm);
+
+                // Remove user3
+                new UserManager(sessionUR).removeUser(fooRealm, user3);
+
+                // Cleanup
+                realmMgr = new RealmManager(sessionUR);
+                realmMgr.removeRealm(realmMgr.getRealm("foo"));
+            }
+        });
+    }
+
+    @Test
+    @ModelTest
+    public void testExpired(KeycloakSession session) {
```

Comment:
The "testExpired" test was updated in the master due some offlineSessions preloading work. Could you try to rebase and doublecheck that new bits added to the test in the meantime are reflected in the migrated test too?